src/celengine/astrooctree.cpp

```diff
@@ -0,0 +1,529 @@
+//     fmt::fprintf(cout, "Adding to node %i object %i with mag %f.\n", this, obj, obj->getAbsoluteMagnitude());
+    if (obj == nullptr)
+    {
+        cerr << "Octree: trying to add null object!!\n";
+        return false;
+    }
+    m_objects.insert(obj);
+    obj->setOctreeNode(this);
+    setDirty(dirty);
+    return true;
+}
+
+bool OctreeNode::rm(LuminousObject *obj)
+{
+//     fmt::fprintf(cout, "OctreeNode::rm(%i (%f))\n", obj->getIndex(), obj->getAbsoluteMagnitude());
+    m_objects.remove(obj);
+    obj->setOctreeNode(nullptr);
+    setDirty(true);
+    return true;
+}
+
+OctreeNode *OctreeNode::getChild(int i, bool create)
+{
+//     if (!Children::isValid(i))
+//         return nullptr;
+    if (m_children == nullptr && create)
+        m_children = new Children;
+    if (m_children != nullptr)
+        return m_children->getChild(i, create ? this : nullptr);
+    return nullptr;
+}
+
+bool OctreeNode::deleteChild(int n)
+{
+//     if (!childValid(n))
+//         return false;
+//     if (m_children->at(n) == nullptr)
+//         return false;
+    m_children->deleteChild(n);
+    if (m_children->getChildrenCount() == 0)
+    {
+        delete m_children;
+        m_children = nullptr;
+    }
+    return true;
+}
+
+bool OctreeNode::insertObject(LuminousObject *obj)
+{
+//     fmt::fprintf(cout, "OctreeNode::insertObject(%i (%f))\n", obj->getIndex(), obj->getAbsoluteMagnitude());
+//     m_objects.dump();
+    OctreeNode *node = this;
+    while (!node->isInCell(obj->getPosition()))
+    {
+        if (getParent() != nullptr)
+        {
+            node = node->getParent();
+        }
+        else
+            return false;
+    }
+    float mag = obj->getAbsoluteMagnitude();
+    while(node->getBrightest() > mag && node->getParent() != nullptr)
+    {
+        node = node->getParent();
+    }
+    while(node->getFaintest() < mag && node->m_objects.size() == node->m_maxObjCount)
+    {
+        node = node->getChild(node->getChildId(obj->getPosition()));
+    }
+    node->add(obj);
+    while(node->m_objects.size() > node->m_maxObjCount)
+    {
+        LuminousObject *obj = node->popFaintest();
+        node = node->getChild(node->getChildId(obj->getPosition()));
+        node->add(obj);
+    }
+    return true;
+}
+
+bool OctreeNode::removeObject(LuminousObject *obj)
+{
+//     fmt::fprintf(cout, "OctreeNode::removeObject(%i (%f))\n", obj->getIndex(), obj->getAbsoluteMagnitude());
+    if (!rm(obj))
+    {
+        fmt::fprintf(clog, "Object %i not found to remove!\n", obj);
+        return false;
+    }
+    OctreeNode *node = this;
+    while(node->m_objects.size() < node->m_maxObjCount && node->getChildrenCount() > 0)
+    {
+//         fmt::fprintf(cout, "Node %i[%i/ %i]->%i\n", node, node->m_objects.size(), node->m_maxObjCount, node->getChildrenCount());
+        int i = node->getBrightestChildId();
+        if (i < 0)
+        {
+            fmt::fprintf(clog, "No brightest child of node %i!!\n", node);
```

Comment:
`node` is a pointer. why do you use `%i`?

Reply:
Indeed. I just didn't pay much attention.